### DISCUSSION TITLE
I try to create react library with Webpack.

### DISCUSSION THREAD

**Discussion by rosinfotech** [webpack/webpack#14303] [Q&A]
# Bug report

**What is the current behavior?**

Modules that were previously split and then built are not loaded.

**If the current behavior is a bug, please provide the steps to reproduce.**

I try to create react library with Webpack.

```
+- Base
|
|--- index.js

export * from "./Base";

|--- Base.jsx

export const Base = { "base_property_1": "base_property_1_value" };

+- Data
|
|--- index.js

export * from "./Data";

|--- Data.jsx

import { Base } from "../Base";

export const Data = { "data_property_1": "data_property_1_value", ...Base };

+- Red
|
|--- index.js

export * from "./Red";

|--- Red.jsx

import React from "react";

export const Red = () => <div>[Red]</div>;
```

I try to build a library with this webpack.library_create.config.js:

```
const path = require("path");
const webpack = require("webpack");

module.exports = {
    entry: {
        "Base": path.resolve(__dirname, "./library_src/Base"),
        "Data": path.resolve(__dirname, "./library_src/Data"),
        "Red": path.resolve(__dirname, "./library_src/Red"),
    },
    externals: {
        "react": "commonjs react",
        "react-dom": "commonjs react-dom",
    },    
    module: {
        rules: [
            {
                test: /\.(js|jsx)$/,
                exclude: /node_modules/,
                use: ["babel-loader"],
            },
        ],
    },
    optimization: {
        minimize: false,
        splitChunks: {
            chunks: "all",
            minChunks: 2,
            minSize: 1,
        },
    },
    output: {
        clean: true,
        filename: "[name]/index.js",
        libraryTarget: "umd",
        library: "Astra",
        path: path.resolve(__dirname, "./node_modules/@a/library"),
        umdNamedDefine: true,
    },
    resolve: {
        extensions: ["*", ".js", ".jsx"],
    },
    target: "web",
};
```

Then I try to start project:

```
+- dist
|
|--- index.html

<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=0" />
    </head>
    <body>
        <div></div>
        <script src="./bundle.js"></script>
    </body>
</html>


+- library_use
|
|--- index.jsx

import React from "react";
import ReactDOM from "react-dom";
import { Data } from "@a/library/Data";
import { Red } from "@a/library/Red";

console.log(Data);

const App = () => <div>App <Red /></div>;

ReactDOM.render( <App />, document.querySelectorAll( "body > div" )[ 0 ] );
```

using webpack.library_use.config.js:

```
const path = require("path");
const webpack = require("webpack");

module.exports = {
    entry: path.resolve(__dirname, "./library_use/index.jsx"),
    module: {
        rules: [
            {
                test: /\.(js|jsx)$/,
                exclude: /node_modules/,
                use: ["babel-loader"],
            },
        ],
    },
    resolve: {
        extensions: ["*", ".js", ".jsx"],
    },
    output: {
        path: path.resolve(__dirname, "./dist"),
        filename: "bundle.js",
    },
    plugins: [ new webpack.HotModuleReplacementPlugin() ],
    devServer: {
        hot: true,
        open: true,
        port: 3020,
        static: [
            path.resolve(__dirname, "./dist"),
            {
                directory: path.resolve(__dirname, "./dist"),
                serveIndex: true,
                watch: true,
            },
        ],        
    },
};
```

And I got an error: Uncaught TypeError: Cannot read properties of undefined (reading 'Data')

i.e. Data is not initialized.

If I remove dependency Base from Data, then Data initialized:

```
+- Data
|
|--- Data.js

// import { Base } from "../Base";

// export const Data = { "data_property_1": "data_property_1_value", ...Base };
export const Data = { "data_property_1": "data_property_1_value", };
```

For convenience, I created a repo to reproduce this bug https://github.com/rosinfotech/issue_210921_webpack_library_dependencies

**What is the expected behavior?**

I expect Webpack have to load splitted dependencies

**Other relevant information:**

webpack version: 5.53.0,
Node.js version: v15.6.0
Operating System:  macOS Big Sur 11.4
Additional tools: –

**Comment by alexander-akait** [comment]
What do you mean splitted?

**Reply by rosinfotech** [reply]
I mean

```
    entry: {
        "Base": path.resolve(__dirname, "./library_src/Base"),
        "Data": path.resolve(__dirname, "./library_src/Data"),
        "Red": path.resolve(__dirname, "./library_src/Red"),
    },
 ```

  and 

```
        splitChunks: {
            chunks: "all",
            minChunks: 2,
            minSize: 1,
        },
```

As result I have 3 directories with 3 files, included packed file Data/index.js related (via import) with packed file Base/index.js, but Base/index.js do not load when I import Data/index.js.

**Reply by alexander-akait** [reply]
Expected, when you split initial chunks you need to load them manually in html/js, otherwise it is not initial chunks

**Reply by rosinfotech** [reply]
But inside code Data/index.js have links with Base/index.js. Besides this, variable Data is not initialized if there is import Base/index, and is initialized otherwise.

I.e. in this case after packing I will have an error Uncaught TypeError: Cannot read properties of undefined (reading 'Data')

```
|- Data.js

import { Base } from "../Base";
export const Data = { "data_property_1": "data_property_1_value", ...Base };
```

And in this case, will not:

```
|- Data.js

export const Data = { "data_property_1": "data_property_1_value" };
```

**Reply by alexander-akait** [reply]
Please look at `dependOn` https://webpack.js.org/configuration/entry-context/#entry-descriptor

**Reply by rosinfotech** [reply]
Same result (not working):

```
    entry: {
        "Base": path.resolve(__dirname, "./library_src/Base"),
        "Data": {
            dependOn: "Base",
            import: path.resolve(__dirname, "./library_src/Data"),
        },
        "Red": path.resolve(__dirname, "./library_src/Red"),
    },
```

```
    // #10 Not working: Data undefined
    output: {
        clean: true,
        filename: "[name]/index.js",
        libraryTarget: "umd",
        library: "Astra",
        path: path.resolve(__dirname, "./node_modules/@a/library"),
        umdNamedDefine: true,
    },
```

Webpack not working, when we build the app with builds that contain dependencies. Or I have a mistake in my config. Which is unlikely.

https://github.com/rosinfotech/issue_210921_webpack_library_dependencies/tree/dependOn
npm run library_create && npm run library_use

**Reply by alexander-akait** [reply]
Look at generated code and you found problems with configuration

**Reply by rosinfotech** [reply]
I tried to debug the code generated by Webpack. And I understood this code requires much time to figure out and resolve this ordinary task. That is why I decided to, that this issue collaborators resolve faster, and I published the Webpack issue.

Never before have I worked with Rollup. Yesterday I spent about 4 hours and found the solution to this task with Rollup https://github.com/rosinfotech/issue_210921_webpack_library_dependencies/tree/rollup

I am really sorry that Webpack could not propose a solution. But maybe someone will see this problem and propose a Webpack solution rely on their own experience. Or maybe this is a Webpack bug and it will fix. I hope.
